README.md
# vstime README

vstime is a VS Code Extension used to provide some simplistic time tracking as you work inside Visual Studio Code
The goal is to make time tracking accessible, right within your editor.
The extension is best for people who spends a lot of time in Visual Studio Code

## Features

* Time Tracking with notes
* Tracking of the current ongoing task
* Ability to Pause
* Detailed breakdown of time spent in different files
* Report of usage

![Demo](images/demo.gif)

## Configuration
* By going to Settings > Extensions > VS Time, you will see a bunch of configuration
* Available configuration
  * Logfull: On by default, meaning verbose logging including how much time you spent in each file is tracked.  

## How to use
* Ctrl + Shift + P and use the following commands
* Start Tracking
* Pause Tracking
  * You can also pause by clicking the status bar item 
* Resume Tracking
  * You can also resume by clicking the status bar item 
* End Tracking
* Open Logs
  * You can view the raw log from here, the logs are stored at your vscode directory
    * For example, on linux it would be ~/.config/Code/User/globalStorage/swuvsmarket.vstime
* Reset Logs
  * Makes a copy of the current log and clear
  
![How to](images/howto.gif)

## Why do I want to try it
Tracking time helps you by:
* Keeping you focused on a single task
* Providing a trail of work habit
* Generating a report for your standup meetings
## Known Issues
* Records are global, which means they are not workspace specific
* Currently there is no roll over, the log file grows indefinitely

CHANGELOG.md
# Change Log

All notable changes to the "vstime" extension will be documented in this file.

## Released
## 0.2.0
- Fixed up a reset issue where previous time logs got carried over

## 0.3.2
- Adding functionality to open the timeing information file via vstime: Open and ability to clear log file via vstime: Reset

## 0.3.7
- Adding a yesterday panel to show a report for one day prior

## 0.3.9
- Fixed an issue with refresh, refresh will refresh all panels

## 0.4.0
- Adding an UI button to stop timer
- Adding a configuration option to turn off verbose tracking
  
  
## [Not released]
### 0.0.2
- Updating meta data (images)
### 0.0.1
- Initial release

src/test/suite/reports.test.ts
import * as assert from 'assert';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
import { Report } from '../../implementations/reporting';
import { TimeTrackingResultItem, TrackerValue } from '../../models/trackerValues';
// import * as myExtension from '../../extension';

suite('Reports Test Suite', () => {
	vscode.window.showInformationMessage('Start report tests.');

	test('Group By', () => {
		const resultItems: TimeTrackingResultItem[] = [
			{ date: 5, comment: 'c', notes: 'c1', logs: [], breakdowns: [], total: new TrackerValue()},
			{ date: 5, comment: 'd', notes: 'd1', logs: [], breakdowns: [], total: new TrackerValue()},
			{ date: 1, comment: 'a', notes: 'a1', logs: [], breakdowns: [], total: new TrackerValue()},
		];
		const report = new Report(resultItems);
		const grouped = report.groupByReport();

		assert.strictEqual(grouped[5].length, 2);
		assert.strictEqual(grouped[1].length, 1);
	});

	test('Report Text', ()=>{
		const total1 = new TrackerValue();
		total1.total = 2000;
		const total2 = new TrackerValue();
		total2.total = 1034;

		const testDate = new Date(2020,10,1);
		const earlierDate = new Date(2020, 9, 24);

		const resultItems: TimeTrackingResultItem[] = [
			{ date: testDate.getTime(), comment: 'Nov1', notes: 'c1', logs: [], breakdowns: [], total: total1},
			{ date: testDate.getTime(), comment: 'Nov2', notes: 'd1', logs: [], breakdowns: [], total: total2},
			{ date: earlierDate.getTime(), comment: 'Oct1', notes: 'a1', logs: [], breakdowns: [], total: total1},
		];
		const report = new Report(resultItems);
		const value =report.report();
		assert.strictEqual(1, 1);
	});
});


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { Tracker } from './implementations/implementation';
import * as initilizers from './initializers/initializer';

let tracker : Tracker;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "vstime" is now active!');
	tracker = initilizers.initialize(context);
}

// this method is called when your extension is deactivated
export function deactivate() {
	if (tracker){
		tracker.stopTracker().then(()=>{});
	}
}


src/models/trackerValues.ts
import { TreeItem, TreeItemCollapsibleState } from "vscode";
import { formatTime } from "../implementations/implementation";

export class TrackerValue{
    total: number;
    lastTrackedAt: number;

    constructor(){
        this.total = -1;
        this.lastTrackedAt = Date.now();
    }

    update() : number {
        const now = Date.now();
        const difference = now - this.lastTrackedAt;
        this.total += difference;
        this.lastTrackedAt = now;

        return this.total;
    }
    
    resume(){
        this.lastTrackedAt = Date.now();
    }

    reset(){
        this.total = -1;
        this.lastTrackedAt = -1;
    }
};


export interface TimeTrackingResultItem{
    date: number,
    comment: string | undefined,
    notes: string | undefined,
    total: TrackerValue,
    breakdowns: {key: string, value: TrackerValue}[]
    logs: string[]
}

export class TimeTrackingItem extends TreeItem implements IHiearchicalNode{
    private date: number;
    private value: TrackerValue;
    private breakdowns: TimeTrackingBreakdownItem[];
    
    children: TreeItem[] = [];

    get breakdown(): TimeTrackingBreakdownItem[] {
        return this.breakdowns;
    }

    constructor(item: TimeTrackingResultItem)
    {
        super('', TreeItemCollapsibleState.Collapsed);
        this.date = item.date;
        this.value = item.total;

        this.breakdowns = item.breakdowns.map(c =>{
            return new TimeTrackingBreakdownItem(c.key, c.value);
        });
        
        const defaultComment = new Date(item.date).toLocaleDateString();
        this.label = item.comment ? item.comment === '' ? defaultComment : item.comment
            : defaultComment;
        this.children = [this.summary(item), ...this.breakdowns];
    }

    summary(item: TimeTrackingResultItem) : TimeTrackingSummaryItem {
        return new TimeTrackingSummaryItem(item.comment ?? "Activity", item.total);
    }
}

export class TimeTrackingBreakdownItem extends TreeItem{
    private value: TrackerValue;

    constructor(detail: string, value: TrackerValue){
        super(detail);
        this.value = value;

        this.label = `${detail}: ${formatTime(value.total)}`;
        this.contextValue = 'breakdown';
    }
}

export class TimeTrackingSummaryItem extends TreeItem{
    constructor(title: string, value: TrackerValue){
        super(title, TreeItemCollapsibleState.None);

        this.contextValue = 'summary';
        this.label = `Total: ${formatTime(value.total)}`;
    }
}

export interface IHiearchicalNode{
    children: TreeItem[]
}


src/initializers/initializer.ts
import * as vscode from "vscode";
import * as impl from "../implementations/implementation";
import * as provider from "../implementations/trackeritemTreeProvider";
import { TimeTrackingResultItem } from "../models/trackerValues";
import * as reporting from "../implementations/reporting";

export function initialize(context: vscode.ExtensionContext): impl.Tracker {
  // register where we can keep our data
  impl.setStoragePath(context.globalStorageUri.fsPath);

  const tracker = initializeTracker(context);

  initializeTree(context, tracker);
  initializeContextCommands(context);

  return tracker;
}
const dayFilter = (t: TimeTrackingResultItem, days: number): boolean => {
  const today = impl.getToday();
  const todayVal = today.getTime();
  const targetDay = impl.getToday();
  targetDay.setDate(today.getDate() - days);

  const startDayVal = targetDay.getTime();
  return t.date >= startDayVal && t.date < todayVal;
};

const weeklyFilter = (t: TimeTrackingResultItem): boolean => {
  return dayFilter(t, 7);
};

const yesterdayFilter = (t: TimeTrackingResultItem): boolean => {
  return dayFilter(t, 1);
};

function initializeTree(
  context: vscode.ExtensionContext,
  tracker: impl.Tracker
) {
  const treeprovider = new provider.TrackerItemTreeProvider(tracker);
  const todayProvider = vscode.window.registerTreeDataProvider(
    "today",
    treeprovider
  );

  const weeklyTreeProvider = new provider.TrackerItemTreeProvider(
    new impl.EmptyTracker(),
    weeklyFilter
  );
  const weekProvider = vscode.window.registerTreeDataProvider(
    "thisweek",
    weeklyTreeProvider
  );

  const yesterdayTreeProvider = new provider.TrackerItemTreeProvider(
    new impl.EmptyTracker(),
    yesterdayFilter
  );
  const yesterdayProvider = vscode.window.registerTreeDataProvider(
    "yesterday",
    yesterdayTreeProvider
  );

  const refreshCommand = vscode.commands.registerCommand(
    "vstime.refresh",
    (c) => {
      treeprovider.refresh();
      weeklyTreeProvider.refresh();
      yesterdayTreeProvider.refresh();
    }
  );

  context.subscriptions.push(todayProvider);
  context.subscriptions.push(weekProvider);
  context.subscriptions.push(yesterdayProvider);

  context.subscriptions.push(refreshCommand);
}

function initializeContextCommands(context: vscode.ExtensionContext) {
  const exportcmd = vscode.commands.registerCommand("vstime.export", (i) => {
    reporting.getReports();
  });
  const opencmd = vscode.commands.registerCommand("vstime.open", (i) => {
    impl.open();
  });
  const resetcmd = vscode.commands.registerCommand("vstime.reset", (i) => {
    impl.reset();
  });

  context.subscriptions.push(exportcmd);
  context.subscriptions.push(opencmd);
  context.subscriptions.push(resetcmd);
}

function initializeTracker(context: vscode.ExtensionContext): impl.Tracker {
  const [statusBar, stopBarItem] = initializeStatusBar(context);
  const tracker = new impl.Tracker(statusBar, stopBarItem);

  let trackStart = vscode.commands.registerCommand("vstime.start", (p) => {
    tracker.startTracker();
  });
  let trackStop = vscode.commands.registerCommand("vstime.stop", async () => {
    const results = await tracker.stopTracker();
    await impl.save(results);
    tracker.reset();
  });

  let trackResume = vscode.commands.registerCommand("vstime.resume", () => {
    tracker.resumeTracker();
  });
  let trackPause = vscode.commands.registerCommand("vstime.pause", () => {
    tracker.pauseTracker();
  });

  context.subscriptions.push(trackStart);
  context.subscriptions.push(trackStop);
  context.subscriptions.push(trackResume);
  context.subscriptions.push(trackPause);

  // Set up when user switches tabs on vscode
  context.subscriptions.push(
    vscode.window.onDidChangeActiveTextEditor((editor) => {
      if (editor && editor === vscode.window.activeTextEditor) {
        tracker.trackChanges(editor?.document.uri.path ?? "Untitled");
      }
    })
  );

  return tracker;
}

function initializeStatusBar(
  context: vscode.ExtensionContext
): vscode.StatusBarItem[] {
  const statusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Left,
    3
  );
  statusBarItem.text = "Timer Off";
  statusBarItem.command = "vstime.start";
  statusBarItem.color = "#BEDEAD";
  context.subscriptions.push(statusBarItem);
  statusBarItem.show();

  const stopBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Left, 2
  );
  stopBarItem.text = "\u25A0\u25A0";
  stopBarItem.command = "vstime.stop";
  stopBarItem.color = "#DEADBE";
  context.subscriptions.push(stopBarItem);
  stopBarItem.show();

  return [statusBarItem, stopBarItem];
}


src/implementations/reporting.ts
import * as vscode from 'vscode';
import { TimeTrackingResultItem } from '../models/trackerValues';
import { formatTime, load, Tracker } from './implementation';

class DateValue{
    date: Date;
    total: number;

    constructor(){
        this.date = new Date();
        this.total = -1;
    }
}
export class Report{
    private results: TimeTrackingResultItem[];

    constructor (results: TimeTrackingResultItem[]){
        this.results = results;
    }

    report(): string{
        // What a raw report could look like
        // Sunday, Nov 01
        // Total: 03:20:34
        // Comment 1: 01:20:32
        //    Spent {time} in {file}
        //    Notes: blah
        // Comment 2: 03:20:40
        //    Spent {time} in {file}
        //    Notes: blah
        let report = '----------------------------------------\r\n';
        const items = this.groupByReport();
        for (var index in items){
            const t = items[index];
            const total:DateValue = t.reduce<DateValue>((accum, c, i, array) =>{
                accum.date = new Date(c.date);
                accum.total += c.total.total;
                return accum;
            }, new DateValue());

            report += `${total.date.toDateString()}\nTotal: ${formatTime(total.total)}\n`;

            t.forEach(detail =>{
                report += `${formatTime(detail.total.total)}: ${detail.comment}\n`;
                  detail.breakdowns.forEach(b =>{
                       report += `  Spent ${formatTime(b.value.total)} in ${b.key}\n`;
                  });
                
                if (detail.notes){
                    report += `    Notes: ${detail.notes}\n\n`;
                }
            });
        }

        report += '----------------------------------------\r\n';
        return report;
    }

    groupByReport() : TimeTrackingResultItem[][] {
        // let report = '----------------------------------------';
        return this.results.sort((a, b) => a.date > b.date ? 1 : -1).reverse()
        .reduce((accum: TimeTrackingResultItem[][], current)=>
        {
            if (accum[current.date]){
                accum[current.date] = [...accum[current.date], current];
            }
            else {
                accum[current.date] = [current];
            }

            return accum;
        }, []);
    }
}

export async function getReports(){
    const results = await load();
    const reportContent = new Report(results).report();
    vscode.workspace.openTextDocument({content: reportContent}).then((doc)=> vscode.window.showTextDocument(doc));
}

src/implementations/implementation.ts
import * as vscode from "vscode";
import * as fs from "fs";
import * as path from "path";
import { TrackerValue, TimeTrackingResultItem } from "../models/trackerValues";

interface KeyNumberPair {
  [key: string]: TrackerValue;
}

const filePath = function(){
  return storagePath;
};

let storagePath: string = '';

export async function test () {
    console.log(vscode.env.appName);
    console.log(vscode.env.appRoot);
    const config = vscode.workspace.getConfiguration("vstime");
    const filename = config.get("filename");
    
    fs.writeFileSync(filePath(), "[]");
}

export function setStoragePath(contextPath: string){
  const targetPath = path.join(contextPath,'.vstime');
  console.log(targetPath);
  if (!fs.existsSync(path.dirname(targetPath))){ fs.mkdirSync(path.dirname(targetPath));}
  storagePath = targetPath;
}

export async function save(timeTrackingDetail: TimeTrackingResultItem) {
  const activitiesTracked = await load();
  const full = [...activitiesTracked, timeTrackingDetail];

  await fs.writeFile(filePath(), JSON.stringify(full), null, (()=>{}));
}

export async function load(): Promise<TimeTrackingResultItem[]> {
  const file = filePath();
  if (fs.existsSync(file))
  {
    const content = fs.readFileSync(file, "utf8");
    return JSON.parse(content === '' ? '[]' : content);
  }
  return [];
}

export async function open() {
    vscode.window.showTextDocument(vscode.Uri.file(storagePath));
}

export async function reset() {
  if (fs.existsSync(storagePath)){
    fs.copyFileSync(storagePath, storagePath + new Date().getTime());
    fs.writeFileSync(storagePath, '');
  }
}

export class Tracker {
  private statusItem: vscode.StatusBarItem;
  private stopStatusItem: vscode.StatusBarItem;
  private timerId: NodeJS.Timeout | null;
  private current: TrackerValue;
  private trackedFiles: KeyNumberPair;
  private comment: string | undefined;
  private currentFile: string;
  private logs: string[];

  isTracking: boolean;
  private logFull: boolean;

  constructor(statusItem: vscode.StatusBarItem,
    stopStatusItem: vscode.StatusBarItem) {

    this.stopStatusItem = stopStatusItem;
    this.statusItem = statusItem;
    this.timerId = null;
    this.currentFile = "";
    this.trackedFiles = {};
    this.current = new TrackerValue();
    this.logs = [];

    this.isTracking = false;

    const config = vscode.workspace.getConfiguration("vstime");
    this.logFull = config.get("logfull") ?? true;

    stopStatusItem.hide();
  }

  get currentProgress() : TimeTrackingResultItem[]{
    if (!this.isTracking) {
      return [];
    }

    return [
       {
        date: getToday().getTime(),
        comment: (!this.comment || this.comment === '') ? 'Current' : `Current: ${this.comment}`,
        notes: '',
        total: this.current,
        breakdowns: this.getBreakdownInfo(),
        logs: []
      }
    ];
  }

  async startTracker() {
    this.comment = await vscode.window.showInputBox({
      prompt: "What are you working on?",
    });
    this.reset();
    this.start();
    this.trackCurrentFile();

    this.stopStatusItem.show();
  }

  resumeTracker() {
    this.logs.push(`resumed at : ${Date.now()}`);
    const currentFile = vscode.window.activeTextEditor?.document.uri.path ?? 'Untitled';
    this.trackChanges(currentFile);
    this.start();
  }

  pauseTracker() {
    this.isTracking = false;
    this.statusItem.command = "vstime.resume";
    this.statusItem.text = "Timer Paused";

    this.logs.push(`paused at : ${Date.now()}`);

    this.stopTimer();
  }


  reset(){
      this.isTracking = false;
      this.current = new TrackerValue();
      this.trackedFiles = {};
  }

  async stopTracker(): Promise<TimeTrackingResultItem> {

    this.stopTimer();
    this.stopStatusItem.show();

    this.isTracking = false;
    const finalComment = await vscode.window.showInputBox({
      prompt: "Thoughts, comments, notes",
    });
    this.logs.push(`stopped at : ${Date.now()}`);

    this.statusItem.command = "vstime.start";
    this.statusItem.text = "Timer Off";
    // log to file
    const values = this.logFull ? this.getBreakdownInfo() : [];

    const final: TimeTrackingResultItem = {
      date: getToday().getTime(),
      comment: this.comment,
      notes: finalComment,
      total: this.current,
      breakdowns: values,
      logs: this.logs,
    };

    return final;
  }

  private getBreakdownInfo() {
    return Object.keys(this.trackedFiles)
      .map((k) => {
        const key = path.parse(k).name;
        const extension = path.parse(k).ext;
        return {
          key: `${key}${extension}`,
          value: this.trackedFiles[k]
        };
      })
      .filter(f => f.value.total > 0);
  }

  trackChanges(file: string) {
    if (!this.isTracking) {
      return;
    }

    const lastTracker =
      this.trackedFiles[this.currentFile] ?? new TrackerValue();
    // if deactivating, update lastTime.update
    // if activating, call resume
    this.logs.push(`stopped at ${Date.now()}`);
    lastTracker.update();
    this.trackedFiles[this.currentFile] = lastTracker;

    const tracker = this.trackedFiles[file] ?? new TrackerValue();
    this.trackedFiles[file] = tracker;

    this.logs.push(`working with ${file} at ${Date.now()}`);
    this.currentFile = file;
    tracker.resume();
  }
  
  private trackCurrentFile(){
    this.currentFile = vscode.window.activeTextEditor?.document.uri.path ?? "Untitled";
    this.trackChanges(this.currentFile);
  }
  
  private stopTimer() {
    this.trackCurrentFile();

    if (this.timerId) {
      clearInterval(this.timerId);
    }
  }

  private start() {
    this.isTracking = true;
    this.trackCurrentFile();

    this.current.resume();
    this.statusItem.command = "vstime.pause";

    this.timerId = setInterval(() => {
      const now = Date.now();
      const total = this.current.update();
      this.statusItem.text = `${formatTime(total)}`;
    }, 1000);
  }

}

export class EmptyTracker extends Tracker{
  constructor() {
    super(vscode.window.createStatusBarItem(), vscode.window.createStatusBarItem());
  }
}

  export function getToday() : Date{
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth(), now.getDate());
  }
  
  export function formatTime(seconds: number) {
    const hours = Math.floor(seconds / 60 / 60 / 1000);
    const minutes = Math.floor(seconds / 60 / 1000) - hours * 60;
    const secondsLeft = Math.round(seconds/1000 - hours * 60 * 60 - minutes * 60);

    const hourStr = hours.toString().padStart(2, '0');
    const minutesStr = minutes.toString().padStart(2, '0');
    const secondsStr = secondsLeft.toString().padStart(2, '0');

    return `${hourStr}:${minutesStr}:${secondsStr}`;
  }



src/implementations/trackeritemTreeProvider.ts
import * as vscode from 'vscode';
import * as trackerImpl from './implementation';

import {TrackerValue, TimeTrackingResultItem, TimeTrackingItem, TimeTrackingBreakdownItem} from '../models/trackerValues';

export class TrackerItemTreeProvider implements vscode.TreeDataProvider<vscode.TreeItem>{
    private _onDidChangeTreeData: vscode.EventEmitter<vscode.TreeItem | null> = new vscode.EventEmitter<vscode.TreeItem | null>();
    readonly onDidChangeTreeData: vscode.Event<vscode.TreeItem | null > = this._onDidChangeTreeData.event;

    private tracker: trackerImpl.Tracker;
    private filter: (t: TimeTrackingResultItem) => boolean;

    constructor(tracker: trackerImpl.Tracker, filter?: (t: TimeTrackingResultItem)=>boolean)
    {
        this.tracker = tracker;

        this.filter = filter ? filter : (t: TimeTrackingResultItem)=> {
            const todayValue = trackerImpl.getToday().getTime();
            return t.date === todayValue;
        };
    }

    refresh() {
		this._onDidChangeTreeData.fire(null);
    }
    
    getTreeItem(element: vscode.TreeItem): vscode.TreeItem | Thenable<vscode.TreeItem> {
        return element;
    }

    getChildren(element?: vscode.TreeItem): vscode.ProviderResult<vscode.TreeItem[]> {
        if (element){
            return (element as TimeTrackingItem).children;
        }
        
        return this.getChildrenItem(element);
    }
    
    async getChildrenItem(element?: vscode.TreeItem) : Promise<vscode.TreeItem[]>{
        if (!element){
            const items = 
                await (await trackerImpl.load())
                    .filter((t) => this.filter(t)) ;

            const results: vscode.TreeItem[] = items.map(i => new TimeTrackingItem(i));
            const current: vscode.TreeItem[] = this.tracker.currentProgress.map(i => new TimeTrackingItem(i));

            return [...current,...results];
        }

        const child = element as TimeTrackingItem;
        const results: vscode.TreeItem[] = child.breakdown;
        return results;
    }
}

src/implementations/impelmentation.ts
import * as vscode from "vscode";
import * as fs from "fs";
import * as path from "path";
import { TrackerValue, TimeTrackingResultItem } from "../models/trackerValues";

interface KeyNumberPair {
  [key: string]: TrackerValue;
}

const filePath = function(){
  return storagePath;
};

let storagePath: string = '';

export async function test () {
    console.log(vscode.env.appName);
    console.log(vscode.env.appRoot);
    const config = vscode.workspace.getConfiguration("vstime");
    
    fs.writeFileSync(filePath(), "[]");
}

export function setStoragePath(contextPath: string){
  const targetPath = path.join(contextPath,'.vstime');
  if (!fs.existsSync(path.dirname(targetPath))){ fs.mkdirSync(path.dirname(targetPath));}
  storagePath = targetPath;
}

export async function save(timeTrackingDetail: TimeTrackingResultItem) {
  const activitiesTracked = await load();
  const full = [...activitiesTracked, timeTrackingDetail];

  await fs.writeFile(filePath(), JSON.stringify(full), null, (()=>{}));
}

export async function load(): Promise<TimeTrackingResultItem[]> {
  const file = filePath();
  if (fs.existsSync(file))
  {
    const content = fs.readFileSync(file, "utf8");
    return JSON.parse(content === '' ? '[]' : content);
  }
  return [];
}

export class Tracker {
  private statusItem: vscode.StatusBarItem;
  private timerId: NodeJS.Timeout | null;
  private current: TrackerValue;
  private trackedFiles: KeyNumberPair;
  private comment: string | undefined;
  private currentFile: string;
  private logs: string[];

  isTracking: boolean;

  constructor(statusItem: vscode.StatusBarItem) {
    this.statusItem = statusItem;
    this.timerId = null;
    this.currentFile = "";
    this.trackedFiles = {};
    this.current = new TrackerValue();
    this.logs = [];

    this.isTracking = false;
  }

  get currentProgress() : TimeTrackingResultItem[]{
    if (!this.isTracking) {
      return [];
    }

    return [
       {
        date: getToday().getTime(),
        comment: 'Current',
        notes: '',
        total: this.current,
        breakdowns: this.getBreakdownInfo(),
        logs: []
      }
    ];
  }

  async startTracker() {
    this.comment = await vscode.window.showInputBox({
      prompt: "What are you working on?",
    });
    this.reset();
    this.start();
    this.trackCurrentFile();
  }

  resumeTracker() {
    this.logs.push(`resumed at : ${Date.now()}`);
    const currentFile = vscode.window.activeTextEditor?.document.uri.path ?? 'Untitled';
    this.trackChanges(currentFile);
    this.start();
  }

  pauseTracker() {
    this.isTracking = false;
    this.statusItem.command = "vstime.resume";
    this.statusItem.text = "Timer Paused";

    this.logs.push(`paused at : ${Date.now()}`);

    this.stopTimer();
  }


  reset(){
      this.isTracking = false;
      this.current = new TrackerValue();
  }

  async stopTracker(): Promise<TimeTrackingResultItem> {

    this.stopTimer();

    this.isTracking = false;
    const finalComment = await vscode.window.showInputBox({
      prompt: "Thoughts, comments, notes",
    });
    this.logs.push(`stopped at : ${Date.now()}`);

    this.statusItem.command = "vstime.start";
    this.statusItem.text = "Timer Off";
    // log to file
    const values = this.getBreakdownInfo();

    const final: TimeTrackingResultItem = {
      date: getToday().getTime(),
      comment: this.comment,
      notes: finalComment,
      total: this.current,
      breakdowns: values,
      logs: this.logs,
    };

    console.log(final);

    return final;
  }

  private getBreakdownInfo() {
    return Object.keys(this.trackedFiles)
      .map((k) => {
        const key = path.parse(k).name;
        const extension = path.parse(k).ext;
        return {
          key: `${key}${extension}`,
          value: this.trackedFiles[k]
        };
      })
      .filter(f => f.value.total > 0);
  }

  trackChanges(file: string) {
    if (!this.isTracking) {
      return;
    }

    const lastTracker =
      this.trackedFiles[this.currentFile] ?? new TrackerValue();
    // if deactivating, update lastTime.update
    // if activating, call resume
    this.logs.push(`stopped at ${Date.now()}`);
    lastTracker.update();
    this.trackedFiles[this.currentFile] = lastTracker;

    const tracker = this.trackedFiles[file] ?? new TrackerValue();
    this.trackedFiles[file] = tracker;

    this.logs.push(`working with ${file} at ${Date.now()}`);
    this.currentFile = file;
    tracker.resume();
  }
  
  private trackCurrentFile(){
    this.currentFile = vscode.window.activeTextEditor?.document.uri.path ?? "Untitled";
    this.trackChanges(this.currentFile);
  }
  
  private stopTimer() {
    this.trackCurrentFile();

    if (this.timerId) {
      clearInterval(this.timerId);
    }
  }

  private start() {
    this.isTracking = true;
    this.trackCurrentFile();

    this.current.resume();
    this.statusItem.command = "vstime.pause";

    this.timerId = setInterval(() => {
      const now = Date.now();
      const total = this.current.update();
      this.statusItem.text = `${formatTime(total)}`;
    }, 1000);
  }

}

export class EmptyTracker extends Tracker{
  constructor() {
    super(vscode.window.createStatusBarItem());
  }
}

  export function getToday() : Date{
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth(), now.getDate());
  }
  
  export function formatTime(seconds: number) {
    const hours = Math.floor(seconds / 60 / 60 / 1000);
    const minutes = Math.floor(seconds / 60 / 1000) - hours * 60;
    const secondsLeft = Math.round(seconds/1000 - hours * 60 * 60 - minutes * 60);

    const hourStr = hours.toString().padStart(2, '0');
    const minutesStr = minutes.toString().padStart(2, '0');
    const secondsStr = secondsLeft.toString().padStart(2, '0');

    return `${hourStr}:${minutesStr}:${secondsStr}`;
  }

